KAFKA-7632: Allow fine-grained configuration for compression

dongjinleekr opened the pull request:
This PR implements [KIP-390: Add producer option to adjust compression level](https://cwiki.apache.org/confluence/display/KAFKA/KIP-390%3A+Add+producer+option+to+adjust+compression+level).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dongjinleekr commented:
Please note that this is a draft implementation. It will be updated following the update of zstd-jni by @luben. (see: luben/zstd-jni#80)

bobrik commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -166,6 +166,12 @@
+    public static final String COMPRESSION_LEVEL_CONFIG = "compression.level";
+    private static final String COMPRESSION_LEVEL_DOC = "The compression level for all data generated by the producer. The default level and valid value is up to "
+                                                        + "compression.type. (<code>none</code>, <code>snappy</code>: not available. <code>gzip</code>: 1~9 (default: 1). "
+                                                        + "<code>lz4</code>: 4~7 (default: 4). <code>zstd</code>: 1~22 (default: 3).";
```
What about negative (fast) ones for zstd?

dongjinleekr replied:
I will be updated with zstd-jni-1.3.7-2 accordingly. It was just released about 4 hours from now.

luben replied:
@dongjinleekr , yes, I noticed that you also want it and I pushed a release.

luben commented:
I just shipped zstd-jni-1.3.7-2 that adds methods to query min/max supported compression levels.

dongjinleekr commented:
I just applied zstd-jni-1.3.7-2. Thanks again @luben!!

dongjinleekr commented:
And there is another issue here; as of now, Kafka's checkstyle rule disallows more than 12 parameters for any method. However, with introducing a new configuration entity, `MemoryRecords#builder` method requires 13 parameters - it is why the automated check fails. To complete this PR, this problem should also be solved.

dongjinleekr commented:
And there is another issue here; as of now, Kafka's checkstyle rule disallows more than 13 parameters for any method. However, with introducing a new configuration entity, the constructor of `MemoryRecordsBuilder` (line 86, 160) requires 14 parameters - it is why the automated check fails. To complete this PR, this problem should also be solved.

dongjinleekr commented:
retest this please

dongjinleekr commented:
retest this please

dongjinleekr committed:
Remove MemoryRecordsBuilder(ByteBuffer, byte, CompressionType, TimestampType, long, long, long, short, int, boolean, boolean, int, int): test-only method.

dongjinleekr committed:
a builder to build MemoryRecordsBuilder with a fluent method call, also can support more than 13 parameters.

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, CompressionType, TimestampType, long)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, byte, CompressionType, TimestampType, long, long)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, byte, CompressionType, TimestampType, long)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, byte, CompressionType, TimestampType, long, long, int)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, CompressionType, long, long, short, int, boolean)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, byte, CompressionType, TimestampType, long, long, long, short, int)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, byte, CompressionType, TimestampType, long, long, long, short, int, boolean, int)

dongjinleekr committed:
Remove MemoryRecords#builder(ByteBuffer, byte, CompressionType, TimestampType, long, long, long, short, int, boolean, boolean, int)

dongjinleekr committed:
Refactor KafkaLZ4BlockOutputStream, KafkaLZ4Test + Add support to compression level, block size configuration for lz4.

1. Make KafkaLZ4BlockOutputStream#BLOCKSIZE_[64KB,256KB,1MB,4MB] private.
2. Add KafkaLZ4BlockOutputStream#AVAILABLE_BLOCKSIZES.
3. Remove KafkaLZ4BlockOutputStream(OutputStream, int, boolean), KafkaLZ4BlockOutputStream(OutputStream, int), KafkaLZ4BlockOutputStream(OutputStream): not used.
4. Remove blockChecksum from KafkaLZ4BlockOutputStream(OutputStream, int, boolean, boolean): it is always false. FLG instance is now instantiated with FLG(), which is equivalent to FLG(false). To support block size and compression level, it now receives LZ4Compressor and BD instance as a parameter. Now, KafkaLZ4BlockOutputStream#of creates LZ4Compressor, BD instance with useBrokenFlagDescriptorChecksum parameter as per compression level, block size, and magic. CompressionType.LZ4#wrapForOutput now calls KafkaLZ4BlockOutputStream#of instead of KafkaLZ4BlockOutputStream(OutputStream, boolean); KafkaLZ4BlockOutputStream(OutputStream, boolean) is also removed for it is not used anymore.
5. KafkaLZ4Test: add magic, compressionLevel, blockSize as testing parameters. useBrokenFlagDescriptorChecksum is now determined by magic. blockChecksum is removed.

dongjinleekr committed:
Add GZipOutputStream to support compression level, block size for Gzip

dongjinleekr committed:
Add CompressionConfig + expand CompressionType#wrapForOutput

1. Add CompressionConfig: without this class, the constructor of MemoryRecordsBuilder violates parameter count limit. (it aready has 13 parameters.)
2. MemoryRecordsBuilder now uses CompressionConfig instead of CompressionType.
3. Add level, blockSize parameters to CompressionType#wrapForOutput.
4. Add compressionLevel, compressionBufferSize, compressionConfig to MemoryRecords.Builder.

dongjinleekr committed:
Refactor MemoryRecords#writeEndTransactionalMarker to use MemoryRecords#builder, instead of instantiating MemoryRecordsBuilder directly

dongjinleekr committed:
Refactor MemoryRecords#withRecords to use MemoryRecords#builder, instead of instantiating MemoryRecordsBuilder directly

dongjinleekr committed:
Make broker to use CompressionConfig

1. Make LogValidator#validateMessagesAndAssignOffsets to use CompressionType and CompressionConfig; not CompressionCodec.

dongjinleekr committed:
Add BrokerCompressionTest#testGetTargetCompressionCodec: validates BrokerCompressionCodec#getTargetCompressionCodec(String, CompressionCodec)

dongjinleekr committed:
BrokerCompressionCodec#getTargetCompressionCodec -> BrokerCompressionCodec#getTargetCompressionType. (returns CompressionType)

1. LogAppendInfo, LogValidator now uses CompressionType instead of CompressionCodec.
2. All related tests (BrokerCompressionTest, LogValidatorTest, AbstractFetcherThreadTest) are updated accordingly.

dongjinleekr committed:
Make RecordAccumulator to use CompressionConfig

dongjinleekr commented:
Here is the re-implementation of KIP-390, rebased against the latest trunk. The commits as of present consist like the following:

1. Refactor the existing code.
    1. Replace current `MemoryRecords#builder` implementation with a new one with a fluent method call. It enables `MemoryRecords#builder` to support more than 13 parameters, also making the rebase against the latest trunk easier.
    2. Change `BrokerCompressionCodec#getTargetCompressionCodec` to return `CompressionType` instead of `CompressionCodec`. It enables the introduction of a new class, `CompressionConfig`. (A new test method for validation was also added.)
2. Extend the compression functionality to support compression level and buffer size.
    1. Refactor `KafkaLZ4BlockOutputStream`, `KafkaLZ4Test` and add support for compression level, block size configuration to lz4.
    2. Add `GZipOutputStream` to support compression level, block size for Gzip.
    3. Add `CompressionConfig` which encapsulates compression type, level, and buffer size, and make Producer and Broker use it.
3. Implement two alternative public interfaces that are under discussion:
    - Type A: Adds `compression.level` and `compression.buffer.size`.
    - Type B: Adds `compression.config` (like `listener.security.protocol.map` or `max.connections.per.ip.overrides`)

Since the public interface is not settled yet, this PR currently includes the common commits by step 2. For the final ones, please refer the following branches:

- Type A: https://github.com/dongjinleekr/kafka/tree/feature/KAFKA-7632-a
- Type B: https://github.com/dongjinleekr/kafka/tree/feature/KAFKA-7632-b

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Retest this please.

dongjinleekr commented:
Closes this PR in preference of a new one.
